### PR TITLE
Implemented initial version of the applyAll functionality.

### PR DESCRIPTION
**Pull request by alexanderKoltman**

### CODE REVIEW

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/logic/ApplyLogic.java [hunk 1]
```diff
@@ -35,32 +30,33 @@
...
      */
     public ApplyLogic(@Nonnull DefaultCloudFoundryOperations cfOperations) {
         checkNotNull(cfOperations);
 
-        this.cfOperations = cfOperations;
         this.spaceDevelopersOperations = new SpaceDevelopersOperations(cfOperations);
         this.servicesOperations = new ServicesOperations(cfOperations);
         this.applicationsOperations = new ApplicationsOperations(cfOperations);
+        this.spaceOperations = new SpaceOperations(cfOperations);
         this.spaceDevelopersOperations = new SpaceDevelopersOperations(cfOperations);
```
Comment (yenle2501) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/logic/ApplyLogic.java:57]: redundant call of spaceDevelopersOperations

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/logic/ApplyLogic.java [hunk 2]
```diff
@@ -73,210 +69,119 @@ public void setSpaceOperations(SpaceOperations spaceOperations) {
...
+     * Provides the service of manipulating the state of a cloud foundry instance
+     * such that it matches with a desired configuration ({@link ConfigBean}).
      *
-     * @param desiredApplications the applications that should all be present in the
-     *                            live system after the procedure
-     * @throws ApplyException       if an non-recoverable error occurs during the procedure
-     * @throws NullPointerException if the argument is null
+     * @param desiredConfigBean   desired configuration for a cloud foundry instance
+     * @param loginCommandOptions LoginCommandOptions
+     * @throws NullPointerException if one of the desired parameter is null.
```
Comment (yenle2501) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/logic/ApplyLogic.java:94]:
```suggestion
     * @throws NullPointerException if one of the desired parameters is null.
```

**Comment by cmatgh** [comment]
Overall looks good to me. Fixed an edge case error for the reworked create app method. Should still be tested end to end extensively to make sure bugs are eliminated to a great extend.

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/logic/ApplyLogic.java [hunk 3]
```diff
@@ -84,227 +86,127 @@ public void setSpaceOperations(SpaceOperations spaceOperations) {
...
+            // getting
 
-        Map<String, List<CfChange>> allApplicationChanges = diffResult.getApplicationChanges();
+            ConfigBean liveConfigBean = new ConfigBean();
 
-        if (allApplicationChanges == null || allApplicationChanges.isEmpty()) {
-            log.info("No changes detected, nothing to apply");
-        } else {
-            ApplicationRequestsPlanner requestsPlanner = new ApplicationRequestsPlanner(applicationsOperations);
+            String desiredSpaceName = desiredConfigBean.getTarget().getSpace();
```
Comment (felix-oq) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/logic/ApplyLogic.java:130]: As the desired space might also have been in the `LoginCommandOptions`:
```suggestion
            String desiredSpaceName = targetOperations.getSpace();
```

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/logic/ApplyLogic.java [hunk 4]
```diff
@@ -84,227 +86,127 @@ public void setSpaceOperations(SpaceOperations spaceOperations) {
...
-     * @throws ApplyException       if an non-recoverable error occurs during the
-     *                              procedure
-     * @throws NullPointerException if the argument is null
+     * @param desiredConfigBean   desired configuration for a cloud foundry instance
+     * @param loginCommandOptions LoginCommandOptions
+     * @throws NullPointerException if one of the desired parameters is null.
      */
-    public void applyApplications(@Nonnull Map<String, ApplicationBean> desiredApplications) {
-        checkNotNull(desiredApplications);
+    public void apply(ConfigBean desiredConfigBean, OptionalLoginCommandOptions loginCommandOptions) {
```
Comment (felix-oq) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/logic/ApplyLogic.java:113]: The `LoginCommandOptions` should not be used. Instead use the functions of `TargetOperations` within this method

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/logic/ApplyLogic.java [hunk 5]
```diff
@@ -84,227 +86,127 @@ public void setSpaceOperations(SpaceOperations spaceOperations) {
...
+                log.verbose("Creating space", desiredSpaceName, "completed");
+
+                // switch to desired space
+                log.info("Switching to space", desiredSpaceName);
+
+                // not a good solution but necessary right now, since there is no time accomplish this in
+                // a more elegant way
+                loginCommandOptions.setSpace(desiredSpaceName);
+                initializeOperations(
+                    CfOperationsCreator.createCfOperations(desiredConfigBean.getTarget(), loginCommandOptions));
```
Comment (felix-oq) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/logic/ApplyLogic.java:146]: Why is this necessary?

Comment (cmatgh) [reply cloud.foundry.cli/src/main/java/cloud/foundry/cli/logic/ApplyLogic.java:146]: No longer necessary. Thanks.

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/logic/diff/change/parsing/NewObjectParsingStrategy.java [hunk 6]
```diff
@@ -25,9 +30,31 @@
...
     @Override
     protected List<CfChange> doParse(Change change) {
+        List<CfChange> cfChanges = new LinkedList<>();
+
+        // since JaVers doesn't explicitly create a change object for the space devs when spec bean is a new object
+        // so creating it manually
+        if (change.getAffectedObject().get() instanceof SpecBean) {
+            List<String> spaceDevelopers = ((SpecBean) change.getAffectedObject().get()).getSpaceDevelopers();
+            if (spaceDevelopers != null) {
+                String propertyName = "spaceDevelopers";
```
Comment (felix-oq) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/logic/diff/change/parsing/NewObjectParsingStrategy.java:40]: It would be nicer if the property name was stored in a constant field. Even better if we guarantee that the field name of the `SpecBean` fits the property name via our `ObjectPropertyValidation` class.
